File: 1249_minimum_remove_to_make_valid_paren/min_remove.rb
def min_remove_to_make_valid(s)
    # Time: O(n)
    # Space: O(n)
    stack = []
    str = s.split("")
    str.each_with_index do |char, index|
      if char == '('
          stack.push(index)
      elsif char == ')'
          if !stack.empty?
              stack.pop
          else
              str[index] = ""
          end
      end
    end
    until stack.empty?
        str[stack.pop] = ""
    end
    str.join("")
end

File: 1249_minimum_remove_to_make_valid_paren/min_remove.py
def minRemoveToMakeValid(s: str) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    stack = []
    tmp = list(s)
    for i, char in enumerate(s):
        if char == "(":
            stack.append(i)
        elif char == ")":
            if stack:
                stack.pop()
            else:
                tmp[i] = ""
    for i in stack:
        tmp[i] = ""
    return "".join(tmp)